drivers/crypto/crypto_mtls_shim.c

```diff
@@ -171,13 +222,364 @@ static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
+				    struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_group grp;
+	mbedtls_ecp_point pt;
+	int r;
+
+	ARG_UNUSED(ctx);
+
+	if (pkt->curve != CIPHER_ECC_CURVE_SECP256R1) {
+		return -EINVAL;
+	}
+
+	r = mbedtls_ecp_group_load(&grp, MBEDTLS_ECP_DP_SECP256R1);
+	if (r != 0) {
+		return -EINVAL;
+	}
+
+	mbedtls_ecp_point_init(&pt);
+	r = mbedtls_ecp_point_read_binary(&grp, &pt,
+					  pkt->public_key,
+					  pkt->public_key_len);
+	if (r != 0) {
+		r = mbedtls_ecp_check_pubkey(&grp, &pt);
+	}
+
+	mbedtls_ecp_point_free(&pt);
+	mbedtls_ecp_group_free(&grp);
+
+	return r == 0 ? 0 : -EINVAL;
+}
+
+static int mtls_ecc_gen_shared_key(struct cipher_ctx *ctx,
+				   struct cipher_ecc_pkt *pkt)
+{
+	mbedtls_ecp_keypair keypair;
+	mbedtls_ecp_point Q;
+	mbedtls_mpi z;
+	mbedtls_mpi d;
+	int r;
+
+	ARG_UNUSED(ctx);
+
+	if (pkt->curve != CIPHER_ECC_CURVE_SECP256R1) {
+		return -EINVAL;
+	}
+
+	mbedtls_ecp_keypair_init(&keypair);
+	mbedtls_ecp_group_free(&keypair.grp);
+	r = mbedtls_ecp_group_load(&keypair.grp, MBEDTLS_ECP_DP_SECP256R1);
+	if (r != 0) {
+		goto free_keypair;
+	}
+
+	mbedtls_ecp_point_init(&Q);
+	r = mbedtls_ecp_point_read_binary(&keypair.grp,
+					  &Q, pkt->public_key,
+					  pkt->public_key_len);
+	if (r != 0) {
+		goto free_Q;
+	}
+
+	mbedtls_mpi_init(&z);
+	r = mbedtls_mpi_read_binary(&z, pkt->private_key, pkt->private_key_len);
+	if (r != 0) {
+		goto free_z;
+	}
+
+	mbedtls_mpi_init(&d);
+
+	r = mbedtls_ecdh_compute_shared(&keypair.grp,
+					&z, &Q, &d,
+					mtls_get_entropy_from_device, NULL);
+	if (r != 0) {
+		goto free_d;
+	}
+
+	r = mbedtls_mpi_write_binary(&d, pkt->shared_secret,
+				     pkt->shared_secret_len);
+
+free_d:
+	mbedtls_mpi_free(&d);
+free_z:
+	mbedtls_mpi_free(&z);
+free_Q:
+	mbedtls_ecp_point_free(&Q);
+free_keypair:
+	mbedtls_ecp_keypair_free(&keypair);
+
+	return r == 0 ? 0 : -EIO;
+}
+
+static int mtls_session_setup_ecc(struct cipher_ctx *ctx,
+				  enum cipher_mode mode,
+				  enum cipher_op op)
+{
+	struct mtls_shim_session *data;
+
+	if (mode != CRYPTO_CIPHER_MODE_NONE) {
+		SYS_LOG_ERR("Only NONE mode supported for ECC algo");
+		return -EINVAL;
+	}
+
+	switch (op) {
+	case CRYPTO_CIPHER_OP_ECC_GEN_KEY:
+		ctx->ops.ecc_crypt_hndlr = mtls_ecc_gen_key;
+		break;
+	case CRYPTO_CIPHER_OP_ECC_VALIDATE_PUBKEY:
+		ctx->ops.ecc_crypt_hndlr = mtls_ecc_validate_pubkey;
+		break;
+	case CRYPTO_CIPHER_OP_ECC_GEN_SHARED_KEY:
+		ctx->ops.ecc_crypt_hndlr = mtls_ecc_gen_shared_key;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	data = mtls_alloc_session(CRYPTO_CIPHER_ALGO_ECC);
+	if (!data) {
+		return -ENOSPC;
+	}
+
+	ctx->ops.cipher_mode = mode;
+	ctx->drv_sessn_state = data;
+
+	return 0;
+}
+
+static int mtls_session_setup(struct device *dev,
+			      struct cipher_ctx *ctx,
+			      enum cipher_algo algo,
+			      enum cipher_mode mode,
+			      enum cipher_op op_type)
+{
+	ARG_UNUSED(dev);
+
+	if (ctx->flags & ~(MTLS_SUPPORT)) {
+		SYS_LOG_ERR("Unsupported flag");
+		return -EINVAL;
```

Comment:
Isn't the case of  `-ENOTSUP` ? It seems exactly the same thing of an invalid algorithm (line 572)